File: ui/src/ui/utils/monitor-labels.test.ts
import { describe, expect, it } from "vitest";
import {
  normalizeMonitorLabels,
  parseMonitorLabels,
  validateMonitorLabelsForCluster,
} from "./monitor-labels.ts";

describe("monitor-labels", () => {
  it("parses promql label pairs", () => {
    expect(parseMonitorLabels('job="hadoop-prod",cluster="bj"')).toEqual([
      { key: "job", value: "hadoop-prod" },
      { key: "cluster", value: "bj" },
    ]);
  });

  it("rejects json format", () => {
    expect(() => parseMonitorLabels('{"job":"hadoop-prod"}')).toThrow(/JSON/);
  });

  it("normalizes unquoted values", () => {
    expect(normalizeMonitorLabels("job=hadoop-prod")).toBe('job="hadoop-prod"');
  });

  it("requires labels for active clusters", () => {
    expect(validateMonitorLabelsForCluster("hadoop", "healthy", "").ok).toBe(false);
    expect(validateMonitorLabelsForCluster("hadoop", "inactive", "").ok).toBe(true);
  });

  it("requires domain-specific label keys", () => {
    expect(validateMonitorLabelsForCluster("fi", "healthy", 'env="prod"').ok).toBe(false);
    expect(validateMonitorLabelsForCluster("gbase", "healthy", 'job="gbase-prod"').ok).toBe(
      true,
    );
  });
});

File: src/pkg/ops/monitor_labels_test.go
package ops

import "testing"

func TestParseMonitorLabels(t *testing.T) {
	pairs, err := ParseMonitorLabels(`job="hadoop-prod",cluster="bj-bch-prod"`)
	if err != nil {
		t.Fatal(err)
	}
	if len(pairs) != 2 || pairs[0].Key != "job" || pairs[1].Value != "bj-bch-prod" {
		t.Fatalf("unexpected pairs: %+v", pairs)
	}

	if _, err := ParseMonitorLabels(`{"job":"hadoop-prod"}`); err == nil {
		t.Fatal("expected JSON rejection")
	}
	if _, err := ParseMonitorLabels(""); err != nil {
		t.Fatalf("empty should be nil pairs: %v", err)
	}
}

func TestValidateMonitorLabelsForCluster(t *testing.T) {
	if err := ValidateMonitorLabelsForCluster(DomainHadoop, "inactive", ""); err != nil {
		t.Fatalf("inactive empty: %v", err)
	}
	if err := ValidateMonitorLabelsForCluster(DomainHadoop, "healthy", ""); err == nil {
		t.Fatal("expected required labels for healthy")
	}
	if err := ValidateMonitorLabelsForCluster(DomainHadoop, "healthy", `env="prod"`); err == nil {
		t.Fatal("expected domain key requirement")
	}
	if err := ValidateMonitorLabelsForCluster(DomainHadoop, "healthy", `job="hadoop-prod"`); err != nil {
		t.Fatalf("valid hadoop labels: %v", err)
	}
	if err := ValidateMonitorLabelsForCluster(DomainFI, "warning", `cluster="huhe-fi"`); err != nil {
		t.Fatalf("valid fi labels: %v", err)
	}
	if err := ValidateMonitorLabelsForCluster(DomainGBase, "unknown", `instance="gbase-1"`); err != nil {
		t.Fatalf("valid gbase labels: %v", err)
	}
}

func TestNormalizeMonitorLabels(t *testing.T) {
	got, err := NormalizeMonitorLabels(`job=hadoop-prod, cluster=bj`)
	if err != nil {
		t.Fatal(err)
	}
	want := `job="hadoop-prod",cluster="bj"`
	if got != want {
		t.Fatalf("got %q want %q", got, want)
	}
}
